### PR TITLE
dev/core#2479 Adjust generic copy for localizable fields

### PR DESCRIPTION
**Pull request by samuelsov**
Overview
----------------------------------------
Add multilingual support when copying entities (event, contribution page, price set, profile).

See https://lab.civicrm.org/dev/core/-/issues/2479

Before
----------------------------------------
Example for event:
Peek_26-03-2021_15-25

After
----------------------------------------
All translatable fields are copied based on the localized field of the original entity.

Technical Details
----------------------------------------
The general idea is to:
* loop over the other non current locale,
* change the db locale
* loop over the translatable field to get the original value in the new locale
* persist the updated translatable fields

Comments
----------------------------------------
I'm going to do more tests but it seems to generally work.

### CODE REVIEW

File: CRM/Core/DAO.php [hunk 1]
```diff
@@ -1816,11 +1817,33 @@ public static function copyGeneric($daoName, $criteria, $newData = NULL, $fields
...
+      if (CRM_Core_I18n::isMultilingual() && $localizableFields) {
+        global $dbLocale;
+        $locales = CRM_Core_I18n::getMultilingual();
+        $curLocale = CRM_Core_I18n::getLocale();
+        // loop on other locales
+        foreach ($locales as $locale) {
+          if ($locale != $curLocale) {
+            // setLocale doesn't seems to be reliable to set dbLocale and we only need to change the db locale
+            $dbLocale = '_' . $locale;
+            $newObject->copyLocalizable($object->id, $newObject->id, $fieldsToPrefix, $fieldsToSuffix, $fieldsToReplace);
```
Comment (samuelsov) [inline CRM/Core/DAO.php:1840]: FIXME: $fieldsToPrefix, $fieldsToSuffix, $fieldsToReplace might not be defined.

**Comment by AlainBenbassat** [comment]
@samuelsov Would it also solve this issue? https://lab.civicrm.org/dev/core/-/issues/851

**Comment by samuelsov** [comment]
@AlainBenbassat yes apparently, it uses the same copyGeneric function (https://github.com/civicrm/civicrm-core/blob/8be074be78de0dee8886e3fbf2db1653127750ed/CRM/Event/Form/ManageEvent/EventInfo.php#L240).

The PR still needs work, review and probably unit tests but it should fix any problem with standard copy of translated entities.

**Comment by AlainBenbassat** [comment]
> @AlainBenbassat yes apparently, it uses the same copyGeneric function:
> 
> https://github.com/civicrm/civicrm-core/blob/8be074be78de0dee8886e3fbf2db1653127750ed/CRM/Event/Form/ManageEvent/EventInfo.php#L240
> 
> The PR still needs work, review and probably unit tests but it should fix any problem with standard copy of translated entities.

Cool. Thanks a lot!

**Comment by colemanw** [comment]
@samuelsov are you able to add a unit test?

**Comment by samuelsov** [comment]
@colemanw yes, I think so. I plan to work on this Friday so hopefully, the PR will be ready for review by next week.

**Comment by samuelsov** [comment]
Event is a good example of data with localizable fields that is often copied (using templates or simply by copying an existing content) so I chose this entity for the unit tests. They are far from exhaustive but they work as expected by failing before the PR and passing after.

I think I would like to have a code review at this point.

File: CRM/Core/DAO.php (hunk 1 above)
Comment (colemanw) [reply CRM/Core/DAO.php:1838]: I see you fixed that.

**Comment by colemanw** [comment]
I've looked at the code and don't see any issues with it. Glad to see a unit test was added. I feel like this is merge-ready once the small problem with the test is solved.

**Comment by samuelsov** [comment]
Ok, I'm not able to reproduce the test error on my side. So, instead of comparing the original data to the api result, I now load both the original data and the copy using the api and then compare. Hopefully, this will fix the problem. In any case, I made the test more verbose so it should be easier to debug.

**Comment by samuelsov** [comment]
@colemanw I assume the ball is on your side but just to be sure is there anything else to do on my side ?

**Comment by mlutfy** [comment]
Looks good to me, but will let someone else merge, since same-shop.

**Pull request merged**